test/SlowTests/Client/Subscriptions/ConcurrentSubscriptionsTests.cs

```diff
@@ -818,6 +820,124 @@ public async Task ThrowOnInvalidLicense()
             }
         }
 
+        [Fact]
+        public async Task AfterAckOfBatchShouldSkipModifiedDocIfItDoesNotMatchCriteria()
+        {
+            using (var store = GetDocumentStore())
+            {
+                var collectionSize = 1000;
+                using (var session = store.OpenSession())
+                {
+
+                    for (int i = 0; i < collectionSize; i++)
+                    {
+                        session.Store(new User() { Name = "E" }, $"user/{i}");
+                    }
+
+                    session.SaveChanges();
+                }
+
+                var mre = new AsyncManualResetEvent();
+                var id = store.Subscriptions.Create(new SubscriptionCreationOptions()
+                {
+                    Query = "from Users where Name != 'R'"
+                });
+
+                var count = 0;
+                using (var subscription = store.Subscriptions.GetSubscriptionWorker<User>(new SubscriptionWorkerOptions(id)
+                {
+                    TimeToWaitBeforeConnectionRetry = TimeSpan.FromSeconds(1),
+                    Strategy = SubscriptionOpeningStrategy.Concurrent,
+                    MaxDocsPerBatch = 64
+                }))
+                {
+                    var t = subscription.Run(async x =>
+                    {
+                        using var s = x.OpenAsyncSession();
+                        foreach (var item in x.Items)
+                        {
+                            item.Result.Name = "G";
+                        }
+
+                        await s.SaveChangesAsync();
+
+                        foreach (var item in x.Items)
+                        {
+                            item.Result.Name = "R";
+                        }
+
+                        await s.SaveChangesAsync();
+
+                        var res = Interlocked.Add(ref count, x.NumberOfItemsInBatch);
+                        if (res >= collectionSize)
+                        {
+                            mre.Set();
+                        }
+
+                    });
+
+
+                    Assert.True(await mre.WaitAsync(TimeSpan.FromSeconds(60)));
+
+                    using (var session = store.OpenSession())
+                    {
+                        var u = session.Load<User>("user/1");
+                        Assert.Equal("R", u.Name);
+                    }
+
```

Comment:
BTW if you add here a 1 second delay... you will have nothing in the list

Reply:
changed